scripts/usb_scan.py

```diff
@@ -0,0 +1,618 @@
+    _KEY_MAP_INTERFACE = {_CLASS: UsbInterface._CLASS,
+                          _SUBCLASS: UsbInterface._SUB_CLASS,
+                          _PROTOCOL: UsbInterface._PROTOCOL}
+
+    _PAT_KEY = r"\s*({}|{}|{}|{}|{}|{})\s*".format(_CLASS, _SUBCLASS,
+                                                   _PROTOCOL, _ID_VENDOR,
+                                                   _ID_PRODUCT, _BCD_DEVICE)
+    _PATTERN = r"{}=\s*([0-9a-f]+)".format(_PAT_KEY)
+
+    _ALLOW = "allow"
+
+    def __init__(self):
+        """ parse policy file, generate rule list
+
+        Note: hubs are never allowed to pass through
+        """
+        self.rule_list = []
+        try:
+            with open(self._PATH, "r") as f:
+                log_wrapper("=== policy file begin")
+                for line in f:
+                    log_wrapper(line[0:-1])
+                    self.parse_line(line)
+                log_wrapper("=== policy file end")
+        except IOError:
+            # without policy file, no device will be allowed to passed through
+            log_wrapper(traceback.format_exc())
+
+        log_wrapper("=== rule list begin")
+        log_list_wrapper(self.rule_list)
+        log_wrapper("=== rule list end")
+
+    def parse_line(self, line):
+        """ parse one line of policy file, generate rule, and append it to
+        self.rule_list
+
+        Example:
+        ALLOW:vid=056a pid=0315 class=03 # Wacom Intuos tablet
+            => {'pid': '0315', 'vid': '056a', 'allow': True, 'class': '03'}
+        DENY: class=03 subclass=01 prot=01 # HID Boot keyboards
+            => {'prot': '01', 'class': '03', 'allow': False, 'subclass': '01'}
+        ALLOW: # Otherwise allow everything else
+            => {'allow': True}
+
+        :param line: (str) single line of policy file
+        :return: None
+        """
+        # 1. remove comments
+        # ^([^#]*)(#.*)?$
+        i = line.find("#")
+        if i != -1:
+            line = line[0:i]
+
+        # 2. split action and match field
+        # ^\s*(ALLOW|DENY)\s*:\s*([^:]*)$
+        if ":" not in line:
+            return
+        parts = line.split(":")
+        n = len(parts)
+        if n == 0 or n > 2:
+            return
+
+        # 3. parse action
+        # \s*(ALLOW|DENY)\s*
+        rule = {}
+        action = parts[0].strip()
+        if action.lower() == "allow":
+            rule[self._ALLOW] = True
+        elif action.lower() == "deny":
+            rule[self._ALLOW] = False
+        else:
+            return
+
+        # 4. parse key=value pairs
+        # pattern = r"\s*(class|subclass|prot|vid|pid|rel)\s*=\s*([0-9a-f]+)"
+        if n == 2:
+            beg, target = 0, parts[1].strip()
+            for matchNum, match in enumerate(re.finditer(self._PATTERN, target,
+                                                         re.IGNORECASE)):
+                if beg != match.start():
+                    log_wrapper("policy line ignored: " + line)
```

Comment:
These should all be logged as errors, as above. At this stage we expect well formed rules.
Clarify the error, something like `"Malformed policy rule, unable to parse '{}'. Ingnored line: {}".format(line[beg,match.start()], line)`

Reply:
add a static function to print detailed info for policy error.